### PR TITLE
Fix termStats posting usage

### PR DESCRIPTION
**Pull request by afoucret**
The current version of the code can cause a buffer overflow because when:

```
if (postingsEnum == null || postingsEnum.advance(docId)) {
  ... do something
}
```

Indeed the call to `advance` need to be done only if `postingsEnum.docID() < targetDocId`

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/script/ScriptTermStatsTests.java [hunk 1]
```diff
@@ -48,9 +48,9 @@ public void testMatchedTermsCount() throws IOException {
 
         // Partial match
         assertAllDocs(
-            Set.of(new Term("field", "foo"), new Term("field", "baz")),
+            Set.of(new Term("field", "foo"), new Term("field", "qux"), new Term("field", "baz")),
```
Comment (afoucret) [inline server/src/test/java/org/elasticsearch/script/ScriptTermStatsTests.java:51]: Added few tests that surface the issue.

File: server/src/main/java/org/elasticsearch/script/ScriptTermStats.java [hunk 2]
```diff
@@ -206,30 +209,26 @@ private TermStates[] loadTermContexts() {
...
+                postings[i] = leafReaderContext.reader().postings(terms[i], PostingsEnum.POSITIONS);
             }
 
             return postings;
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }
     }
+
+    private void advancePostings(int targetDocId) {
```
Comment (afoucret) [inline server/src/main/java/org/elasticsearch/script/ScriptTermStats.java:223]: No having a method that does advance posting to the right id securely.

File: server/src/main/java/org/elasticsearch/script/ScriptTermStats.java [hunk 3]
```diff
@@ -150,8 +150,9 @@ public StatsSummary termFreq() {
         final int docId = docIdSupplier.getAsInt();
 
         try {
+            advancePostings(docId);
             for (PostingsEnum postingsEnum : postingsSupplier.get()) {
-                if (postingsEnum == null || postingsEnum.advance(docId) != docId) {
+                if (postingsEnum == null || postingsEnum.docID() != docId) {
```
Comment (afoucret) [inline server/src/main/java/org/elasticsearch/script/ScriptTermStats.java:155]: Now calling advancePosting before the loop and just check the docID

File: server/src/main/java/org/elasticsearch/script/ScriptTermStats.java [hunk 4]
```diff
@@ -73,8 +72,9 @@ public int matchedTermsCount() {
         int matchedTerms = 0;
 
         try {
+            advancePostings(docId);
             for (PostingsEnum postingsEnum : postingsSupplier.get()) {
-                if (postingsEnum != null && postingsEnum.advance(docId) == docId && postingsEnum.freq() > 0) {
+                if (postingsEnum != null && postingsEnum.docID() == docId && postingsEnum.freq() > 0) {
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/script/ScriptTermStats.java:77]: How would it be possible for `postingsEnum.freq()` to return 0 for a match?

File: server/src/main/java/org/elasticsearch/script/ScriptTermStats.java [hunk 5]
```diff
@@ -206,30 +209,26 @@ private TermStates[] loadTermContexts() {
...
-                if (state == null) {
-                    postings[i] = null;
-                    continue;
-                }
-
-                TermsEnum termsEnum = leafReaderContext.reader().terms(terms[i].field()).iterator();
-                termsEnum.seekExact(terms[i].bytes(), state);
-
-                postings[i] = termsEnum.postings(null, PostingsEnum.ALL);
+                postings[i] = leafReaderContext.reader().postings(terms[i], PostingsEnum.POSITIONS);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/script/ScriptTermStats.java:214]: nice!

**Review by jimczi — APPROVED** [review]
LGTM

**Review by kderusso — APPROVED** [review]

**Review by rjernst — APPROVED** [review]
Ouch. LGTM too.

File: server/src/main/java/org/elasticsearch/script/ScriptTermStats.java (hunk 4 above)
Comment (afoucret) [reply server/src/main/java/org/elasticsearch/script/ScriptTermStats.java]: This is right. Changed.

**Pull request merged**